### PR TITLE
Feature/cdap 13326 refactor schema propagation

### PR DESCRIPTION
**Pull request by albertshau**

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 1]
```diff
@@ -168,11 +168,11 @@ protected void configure() {
     // the run early, before provisioning is performed.
     // If plugins were registered only at the application level, CDAP would not be able to fail the run early.
     try {
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java:170]: please update copyright year

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/spec/SchemaPropagator.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+/**
+ * Keeps track of schemas for various stages of a pipeline and propagates them as they are updated.
+ *
+ * The DefaultPipelineConfigurer's keep state about the input, output, and error schemas for each stage.
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/spec/SchemaPropagator.java]: nit: s/DefaultPipelineConfigurer's/DefaultPipelineConfigurers

Comment (jayfjin) [reply cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/spec/SchemaPropagator.java]: also maybe {@link DefaultPipelineConfigurer}s

**Review by jayfjin — COMMENTED** [review]
Where else is the SchemaPropagator used?

**Comment by albertshau** [comment]
Nowhere else yet, that's coming in a later PR

**Review by jayfjin — APPROVED** [review]

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-DUT6905

**Pull request merged**